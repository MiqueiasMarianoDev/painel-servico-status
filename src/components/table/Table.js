import React, { useMemo } from 'react'
import { useTable } from 'react-table/dist/react-table.development'
import './Table.module.css'
import { Status } from '../status/Status';
import { Loading } from '../loading/Loading';
import { Error } from '../error/Error';


export const Table = ({_columns, _data, loading, error}) => {
  const columns = useMemo(() => _columns, [_columns]);
  const data = useMemo(() => _data, [_data]);

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow
  } = useTable({
    columns,
    data
  })

  return (
    <>
        <table {...getTableProps()}>
            <thead>
            {headerGroups.map(headerGroup => (
                <tr {...headerGroup.getHeaderGroupProps()}>
                {headerGroup.headers.map(column => (
                    <th {...column.getHeaderProps()}>{column.render('Header')}</th>
                ))}
                </tr>
            ))}
            </thead>
            <tbody {...getTableBodyProps()}>
                {rows && rows.map(row => {
                    prepareRow(row)
                    return (
                        <tr {...row.getRowProps()}>
                            {row.cells.map(cell => {
                                if (cell.column.id === "status") {
                                  return <td {...cell.getCellProps()}><Status _status={cell.render('Cell')} /></td>
                                }
                                return <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
                            })}
                        </tr>
                    )
                })
                }
            </tbody>   
        </table>
        {loading && (
          <Loading />
        )}
        {error &&
          <Error />
        }


    </>
  )
}
Table.defaultProps = {
  loading: false,
  error: false
}