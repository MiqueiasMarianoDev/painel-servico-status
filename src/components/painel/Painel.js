import React from 'react'
import { Table } from '../table/Table'
import MOCK_DATA from '../MOCK_DATA.json'
import { COLUMNS } from '../columns'

export const Painel = () => {
  return (
    <div>
      <h2>Painel</h2>
      <h3>Painel com Dados</h3>
      <Table _columns={COLUMNS} _data={MOCK_DATA} />
      <h3>Painel Loading</h3>
      <Table _columns={COLUMNS} _data={[]} />
    </div>
  )
}