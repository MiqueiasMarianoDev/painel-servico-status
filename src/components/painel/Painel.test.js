import React from 'react';
import { render, screen } from '@testing-library/react';
import { Painel } from './Painel';
import { Table } from '../table/Table';

describe('Painel', () => {

  test('renders Table component', () => {
    render(<Painel />);
    const table = screen.getByType(Table);
    expect(table).toBeInTheDocument();
  }); 

  test('renders 3 Table components', () => {
    render(<Painel />);
    const tables = screen.getAllByType(Table);
    expect(tables.length).toBe(3);
  });

  test('first Table has data', () => {
    render(<Painel />);
    const firstTable = screen.getAllByType(Table)[0];
    expect(firstTable).toHaveTextContent('Painel com Dados');
  });

  test('second Table loading=true', () => {
    render(<Painel />);
    const secondTable = screen.getAllByType(Table)[1];
    expect(secondTable.props.loading).toBeTruthy();
  });  

  test('third Table error=true', () => {
    render(<Painel />);
    const thirdTable = screen.getAllByType(Table)[2];
    expect(thirdTable.props.error).toBeTruthy();
  });

  test('snapshot', () => {
    const {asFragment} = render(<Painel />);
    expect(asFragment()).toMatchSnapshot();
  });

});