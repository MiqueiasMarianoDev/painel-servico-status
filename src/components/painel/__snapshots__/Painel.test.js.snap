// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Painel snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Painel
    </h2>
    <h3>
      Painel com Dados
    </h3>
    <table
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
          >
            Id
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            Service
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            1
          </td>
          <td
            role="cell"
          >
            SERPRO
          </td>
          <td
            role="cell"
          >
            <div
              class="status green"
            >
              <svg
                color="green"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: green;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12Zm16.28-2.72a.751.751 0 0 0-.018-1.042.751.751 0 0 0-1.042-.018l-5.97 5.97-2.47-2.47a.751.751 0 0 0-1.042.018.751.751 0 0 0-.018 1.042l3 3a.75.75 0 0 0 1.06 0Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            2
          </td>
          <td
            role="cell"
          >
            Conselho Nacional de Justiça
          </td>
          <td
            role="cell"
          >
            <div
              class="status gray"
            >
              <svg
                color="grey"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: grey;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12Zm11.575-4.75a.825.825 0 1 0-1.65 0v5.5c0 .296.159.57.416.716l3.5 2a.825.825 0 0 0 .818-1.432l-3.084-1.763Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            3
          </td>
          <td
            role="cell"
          >
            Tribunal de Justiça
          </td>
          <td
            role="cell"
          >
            <div
              class="status red"
            >
              <svg
                color="red"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: red;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.036 7.976a.75.75 0 0 0-1.06 1.06L10.939 12l-2.963 2.963a.75.75 0 1 0 1.06 1.06L12 13.06l2.963 2.964a.75.75 0 0 0 1.061-1.06L13.061 12l2.963-2.964a.75.75 0 1 0-1.06-1.06L12 10.939 9.036 7.976Z"
                />
                <path
                  d="M12 1c6.075 0 11 4.925 11 11s-4.925 11-11 11S1 18.075 1 12 5.925 1 12 1ZM2.5 12a9.5 9.5 0 0 0 9.5 9.5 9.5 9.5 0 0 0 9.5-9.5A9.5 9.5 0 0 0 12 2.5 9.5 9.5 0 0 0 2.5 12Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            4
          </td>
          <td
            role="cell"
          >
            Ministério da Saúde
          </td>
          <td
            role="cell"
          >
            <div
              class="status gray"
            >
              <svg
                color="grey"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: grey;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12Zm11.575-4.75a.825.825 0 1 0-1.65 0v5.5c0 .296.159.57.416.716l3.5 2a.825.825 0 0 0 .818-1.432l-3.084-1.763Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            5
          </td>
          <td
            role="cell"
          >
            Receita Federal
          </td>
          <td
            role="cell"
          >
            <div
              class="status green"
            >
              <svg
                color="green"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: green;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12Zm16.28-2.72a.751.751 0 0 0-.018-1.042.751.751 0 0 0-1.042-.018l-5.97 5.97-2.47-2.47a.751.751 0 0 0-1.042.018.751.751 0 0 0-.018 1.042l3 3a.75.75 0 0 0 1.06 0Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            6
          </td>
          <td
            role="cell"
          >
            Departamento de Trânsito
          </td>
          <td
            role="cell"
          >
            <div
              class="status gray"
            >
              <svg
                color="grey"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: grey;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12Zm11.575-4.75a.825.825 0 1 0-1.65 0v5.5c0 .296.159.57.416.716l3.5 2a.825.825 0 0 0 .818-1.432l-3.084-1.763Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            7
          </td>
          <td
            role="cell"
          >
            Banco Central
          </td>
          <td
            role="cell"
          >
            <div
              class="status green"
            >
              <svg
                color="green"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: green;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12Zm16.28-2.72a.751.751 0 0 0-.018-1.042.751.751 0 0 0-1.042-.018l-5.97 5.97-2.47-2.47a.751.751 0 0 0-1.042.018.751.751 0 0 0-.018 1.042l3 3a.75.75 0 0 0 1.06 0Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            8
          </td>
          <td
            role="cell"
          >
            Ministério da Educação
          </td>
          <td
            role="cell"
          >
            <div
              class="status red"
            >
              <svg
                color="red"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: red;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.036 7.976a.75.75 0 0 0-1.06 1.06L10.939 12l-2.963 2.963a.75.75 0 1 0 1.06 1.06L12 13.06l2.963 2.964a.75.75 0 0 0 1.061-1.06L13.061 12l2.963-2.964a.75.75 0 1 0-1.06-1.06L12 10.939 9.036 7.976Z"
                />
                <path
                  d="M12 1c6.075 0 11 4.925 11 11s-4.925 11-11 11S1 18.075 1 12 5.925 1 12 1ZM2.5 12a9.5 9.5 0 0 0 9.5 9.5 9.5 9.5 0 0 0 9.5-9.5A9.5 9.5 0 0 0 12 2.5 9.5 9.5 0 0 0 2.5 12Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            9
          </td>
          <td
            role="cell"
          >
            Agência Nacional de Energia Elétrica
          </td>
          <td
            role="cell"
          >
            <div
              class="status green"
            >
              <svg
                color="green"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: green;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12Zm16.28-2.72a.751.751 0 0 0-.018-1.042.751.751 0 0 0-1.042-.018l-5.97 5.97-2.47-2.47a.751.751 0 0 0-1.042.018.751.751 0 0 0-.018 1.042l3 3a.75.75 0 0 0 1.06 0Z"
                />
              </svg>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            10
          </td>
          <td
            role="cell"
          >
            Instituto Nacional do Seguro Social
          </td>
          <td
            role="cell"
          >
            <div
              class="status gray"
            >
              <svg
                color="grey"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: grey;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12Zm11.575-4.75a.825.825 0 1 0-1.65 0v5.5c0 .296.159.57.416.716l3.5 2a.825.825 0 0 0 .818-1.432l-3.084-1.763Z"
                />
              </svg>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <h3>
      Painel Loading
    </h3>
    <table
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
          >
            Id
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            Service
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      />
    </table>
    <div
      class="loading"
    >
      <div
        class="spinner"
      />
      <p>
        Carregando...
      </p>
    </div>
    <h3>
      Painel Error
    </h3>
    <table
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
          >
            Id
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            Service
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      />
    </table>
    <div
      class="erro"
    >
      <p>
        Oops, ocorreu um erro:
      </p>
      <p />
    </div>
  </div>
</DocumentFragment>
`;
